module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+	entity flow.Entity
+}
+
+type EntityDoubleLinkedList struct {
+	total        uint32
+	free         *doubleLinkedList // keeps track of free slots.
+	used         *doubleLinkedList // keeps track of allocated slots to cachedEntities.
+	values       []cachedEntity
+	ejectionMode EjectionMode
+}
+
+func NewEntityList(limit uint32, ejectionMode EjectionMode) *EntityDoubleLinkedList {
+	l := &EntityDoubleLinkedList{
+		free:         newDoubleLinkedList(),
+		used:         newDoubleLinkedList(),
+		values:       make([]cachedEntity, limit),
+		ejectionMode: ejectionMode,
+	}
+
+	l.initFreeEntities(limit)
+
+	return l
+}
+
+// initFreeEntities initializes the free double linked-list with all slice indices in the range of
+// [0, limit). In other words, all slice indices in cachedEntity are marked as free indices.
+func (e *EntityDoubleLinkedList) initFreeEntities(limit uint32) {
+	e.free.head.setPointer(0)
+	e.free.tail.setPointer(0)
+
+	for i := uint32(1); i < limit; i++ {
+		// appends slice index i to tail of free linked list
+		e.link(e.free.tail, EIndex(i))
+		// and updates its tail
+		e.free.tail.setPointer(EIndex(i))
+	}
+}
+
+// Add writes given entity into a cachedEntity on the underlying entities linked-list. Return value is
+// the index at which given entity is written on entities linked-list so that it can be accessed directly later.
+func (e *EntityDoubleLinkedList) Add(entityId flow.Identifier, entity flow.Entity, owner uint64) EIndex {
+	entityIndex := e.sliceIndexForEntity()
+	e.values[entityIndex].entity = entity
+	e.values[entityIndex].id = entityId
+	e.values[entityIndex].owner = owner
+	e.values[entityIndex].node.next.setUndefined()
+	e.values[entityIndex].node.prev.setUndefined()
+
+	if e.used.head.isUndefined() {
+		// used list is empty, hence setting head of used list to current entityIndex.
+		e.used.head.setPointer(entityIndex)
+		e.values[e.used.head.sliceIndex()].node.prev.setUndefined()
+	}
+
+	if !e.used.tail.isUndefined() {
+		// links new entity to the tail
+		e.link(e.used.tail, entityIndex)
+	}
+
+	// since we are appending to the used list, entityIndex also acts as tail of the list.
+	e.used.tail.setPointer(entityIndex)
+
+	e.total++
+	return entityIndex
+}
+
+// Get returns entity corresponding to the entity index from the underlying list.
+func (e EntityDoubleLinkedList) Get(entityIndex EIndex) (flow.Identifier, flow.Entity, uint64) {
+	return e.values[entityIndex].id, e.values[entityIndex].entity, e.values[entityIndex].owner
+}
+
+// sliceIndexForEntity returns a slice index which hosts the next entity to be added to the list.
+func (e *EntityDoubleLinkedList) sliceIndexForEntity() EIndex {
+	if e.free.head.isUndefined() {
+		// free list is empty, hence, we are out of space, we
+		// need to eject.
+		if e.ejectionMode == RandomEjection {
+			// turning a random entity into a free head.
+			e.invalidateRandomEntity()
+		} else {
+			// LRU ejection
+			// used head is the oldest entity, hence,
```

Comment:
Suggestion: If the ejection mode is a caller supplied function then it could use a more abstract API (abstracting away the complicated double linked lists and fear inducing `head` and `tail` words :-) ) to make the code simpler to understand? For example, this kind of intention revealing API:

```
index := heroPool.GetIndexByOldest(stateUsed)
heroPool.SetStateByIndex(index, stateUsed, stateFree)
```

Reply:
Addressed here: https://github.com/onflow/flow-go/pull/1778/commits/b1ead7d8c6e08392c507ba7d3c9b03d2a1e32f58